app/services/cart_service.rb

```diff
@@ -29,24 +29,32 @@ def attempt_cart_add_variants(variants_data)
 
     variants_data.each do |variant_data|
       loaded_variant = loaded_variants[variant_data[:variant_id].to_i]
+
+      if loaded_variant.deleted?
+        remove_deleted_variant(loaded_variant)
+        next
+      end
+
       next unless varies_from_cart(variant_data, loaded_variant)
 
-      attempt_cart_add(
-        loaded_variant, variant_data[:quantity], variant_data[:max_quantity]
-      )
+      attempt_cart_add(loaded_variant, variant_data[:quantity], variant_data[:max_quantity])
     end
   end
 
   def indexed_variants(variants_data)
     @indexed_variants ||= begin
       variant_ids_in_data = variants_data.map{ |v| v[:variant_id] }
 
-      Spree::Variant.where(id: variant_ids_in_data).
+      Spree::Variant.with_deleted.where(id: variant_ids_in_data).
         includes(:default_price, :stock_items, :product).
         index_by(&:id)
     end
   end
 
+  def remove_deleted_variant(variant)
+    line_item_for_variant(variant).andand.destroy
```

Comment:
I think in the long term we should soft-delete the line item upon soft-deletion of the variant, ala ON DELETE CASCADE only that it'll be an `UPDATE` statement in this case.